no-extra-parens handles precedence incorrectly in some cases

<!--
This template is for bug reports. If you are reporting a bug, please continue on. If you are here for another reason, please see below:

1. To propose a new rule: http://eslint.org/docs/developer-guide/contributing/new-rules
2. To request a change: http://eslint.org/docs/developer-guide/contributing/changes
3. If you have any questions, please stop by our chatroom: https://gitter.im/eslint/eslint

Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

**Tell us about your environment**

* **ESLint Version:** master
* **Node Version:** 7.4.0
* **npm Version:** 4.0.5

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

<!-- Paste your configuration below: -->
```yml
rules:
  no-extra-parens: error
parserOptions:
  ecmaVersion: 2016
```

**What did you do? Please include the actual source code causing the issue.**

<!-- Paste the source code below: -->
```js
(new Foo)();

-(foo ** bar);

(-foo) ** bar;
```

**What did you expect to happen?**

I expected no errors to be reported, because removing parentheses from each of those cases results in a change in behavior or a syntax error.

**What actually happened? Please include the actual, raw output from ESLint.**

3 errors were reported:

```
<text>
  1:1  error  Gratuitous parentheses around expression  no-extra-parens
  3:2  error  Gratuitous parentheses around expression  no-extra-parens
  5:1  error  Gratuitous parentheses around expression  no-extra-parens
```

--- a/lib/ast-utils.js
+++ b/lib/ast-utils.js
@@ -673,6 +673,8 @@ module.exports = {
                     case "/":
                     case "%":
                         return 13;
+                    case "**":
+                        return 15;
 
                     // no default
                 }
@@ -681,10 +683,10 @@ module.exports = {
 
             case "UnaryExpression":
             case "AwaitExpression":
-                return 14;
+                return 16;
 
             case "UpdateExpression":
-                return 15;
+                return 17;
 
             case "CallExpression":
 
@@ -692,14 +694,14 @@ module.exports = {
                 if (node.callee.type === "FunctionExpression") {
                     return -1;
                 }
-                return 16;
+                return 18;
 
             case "NewExpression":
-                return 17;
+                return 19;
 
             // no default
         }
-        return 18;
+        return 20;
     },
 
     /**

--- a/lib/rules/no-extra-parens.js
+++ b/lib/rules/no-extra-parens.js
@@ -168,6 +168,19 @@ module.exports = {
         }
 
         /**
+         * Determines if a constructor function is newed-up with parens
+         * @param {ASTNode} newExpression - The NewExpression node to be checked.
+         * @returns {boolean} True if the constructor is called with parens.
+         * @private
+         */
+        function isNewExpressionWithParens(newExpression) {
+            const lastToken = sourceCode.getLastToken(newExpression);
+            const penultimateToken = sourceCode.getTokenBefore(lastToken);
+
+            return newExpression.arguments.length > 0 || penultimateToken.value === "(" && lastToken.value === ")";
+        }
+
+        /**
          * Determines if a node is or contains an assignment expression
          * @param {ASTNode} node - The node to be checked.
          * @returns {boolean} True if the node is or contains an assignment expression.
@@ -353,6 +366,10 @@ module.exports = {
          * @private
          */
         function dryUnaryUpdate(node) {
+            if (node.type === "UnaryExpression" && node.argument.type === "BinaryExpression" && node.argument.operator === "**") {
+                return;
+            }
+
             if (hasExcessParens(node.argument) && precedence(node.argument) >= precedence(node)) {
                 report(node.argument);
             }
@@ -367,7 +384,8 @@ module.exports = {
         function dryCallNew(node) {
             if (hasExcessParens(node.callee) && precedence(node.callee) >= precedence(node) && !(
                 node.type === "CallExpression" &&
-                node.callee.type === "FunctionExpression" &&
+                (node.callee.type === "FunctionExpression" ||
+                  node.callee.type === "NewExpression" && !isNewExpressionWithParens(node.callee)) &&
 
                 // One set of parentheses are allowed for a function expression
                 !hasDoubleExcessParens(node.callee)
@@ -395,13 +413,17 @@ module.exports = {
          */
         function dryBinaryLogical(node) {
             const prec = precedence(node);
-            const shouldSkipLeft = NESTED_BINARY && (node.left.type === "BinaryExpression" || node.left.type === "LogicalExpression");
+            const leftPrecedence = precedence(node.left);
+            const rightPrecedence = precedence(node.right);
+            const isExponentiation = node.operator === "**";
+            const shouldSkipLeft = (NESTED_BINARY && (node.left.type === "BinaryExpression" || node.left.type === "LogicalExpression")) ||
+              node.left.type === "UnaryExpression" && isExponentiation;
             const shouldSkipRight = NESTED_BINARY && (node.right.type === "BinaryExpression" || node.right.type === "LogicalExpression");
 
-            if (!shouldSkipLeft && hasExcessParens(node.left) && precedence(node.left) >= prec) {
+            if (!shouldSkipLeft && hasExcessParens(node.left) && (leftPrecedence > prec || (leftPrecedence === prec && !isExponentiation))) {
                 report(node.left);
             }
-            if (!shouldSkipRight && hasExcessParens(node.right) && precedence(node.right) > prec) {
+            if (!shouldSkipRight && hasExcessParens(node.right) && (rightPrecedence > prec || (rightPrecedence === prec && isExponentiation))) {
                 report(node.right);
             }
         }

--- a/tests/lib/rules/no-extra-parens.js
+++ b/tests/lib/rules/no-extra-parens.js
@@ -89,6 +89,10 @@ ruleTester.run("no-extra-parens", rule, {
         "(++a)(b); (c++)(d);",
         "new (A())",
         "new A()()",
+        "(new A)()",
+        "(new (Foo || Bar))()",
+        { code: "(2 + 3) ** 4", parserOptions: { ecmaVersion: 7 } },
+        { code: "2 ** (2 + 3)", parserOptions: { ecmaVersion: 7 } },
 
         // same precedence
         "a, b, c",
@@ -120,6 +124,8 @@ ruleTester.run("no-extra-parens", rule, {
         "a(b)(c)",
         "a((b, c))",
         "new new A",
+        { code: "2 ** 3 ** 4", parserOptions: { ecmaVersion: 7 } },
+        { code: "(2 ** 3) ** 4", parserOptions: { ecmaVersion: 7 } },
 
         // constructs that contain expressions
         "if(a);",
@@ -183,6 +189,15 @@ ruleTester.run("no-extra-parens", rule, {
 
         // Object literals as arrow function bodies need parentheses
         { code: "x => ({foo: 1})", parserOptions: { ecmaVersion: 6 } },
+
+
+        // Exponentiation operator `**`
+        { code: "1 + 2 ** 3", parserOptions: { ecmaVersion: 7 } },
+        { code: "1 - 2 ** 3", parserOptions: { ecmaVersion: 7 } },
+        { code: "2 ** -3", parserOptions: { ecmaVersion: 7 } },
+        { code: "(-2) ** 3", parserOptions: { ecmaVersion: 7 } },
+        { code: "(+2) ** 3", parserOptions: { ecmaVersion: 7 } },
+        { code: "+ (2 ** 3)", parserOptions: { ecmaVersion: 7 } },
 
         // https://github.com/eslint/eslint/issues/5789
         { code: "a => ({b: c}[d])", parserOptions: { ecmaVersion: 6 } },
@@ -409,6 +424,11 @@ ruleTester.run("no-extra-parens", rule, {
         invalid("a + (b * c)", "a + b * c", "BinaryExpression"),
         invalid("(a * b) + c", "a * b + c", "BinaryExpression"),
         invalid("(a * b) / c", "a * b / c", "BinaryExpression"),
+        invalid("(2) ** 3 ** 4", "2 ** 3 ** 4", "Literal", null, { parserOptions: { ecmaVersion: 7 } }),
+        invalid("2 ** (3 ** 4)", "2 ** 3 ** 4", "BinaryExpression", null, { parserOptions: { ecmaVersion: 7 } }),
+        invalid("(2 ** 3)", "2 ** 3", "BinaryExpression", null, { parserOptions: { ecmaVersion: 7 } }),
+        invalid("(2 ** 3) + 1", "2 ** 3 + 1", "BinaryExpression", null, { parserOptions: { ecmaVersion: 7 } }),
+        invalid("1 - (2 ** 3)", "1 - 2 ** 3", "BinaryExpression", null, { parserOptions: { ecmaVersion: 7 } }),
 
         invalid("a = (b * c)", "a = b * c", "BinaryExpression", null, { options: ["all", { nestedBinaryExpressions: false }] }),
         invalid("(b * c)", "b * c", "BinaryExpression", null, { options: ["all", { nestedBinaryExpressions: false }] }),
@@ -425,6 +445,11 @@ ruleTester.run("no-extra-parens", rule, {
         invalid("new (\nfunction(){}\n)", "new \nfunction(){}\n", "FunctionExpression", 1),
         invalid("((function foo() {return 1;}))()", "(function foo() {return 1;})()", "FunctionExpression"),
         invalid("((function(){ return bar(); })())", "(function(){ return bar(); })()", "CallExpression"),
+
+        invalid("new (A)", "new A", "Identifier"),
+        invalid("(new A())()", "new A()()", "NewExpression"),
+        invalid("(new A(1))()", "new A(1)()", "NewExpression"),
+        invalid("((new A))()", "(new A)()", "NewExpression"),
 
         invalid("0, (_ => 0)", "0, _ => 0", "ArrowFunctionExpression", 1, { parserOptions: { ecmaVersion: 6 } }),
         invalid("(_ => 0), 0", "_ => 0, 0", "ArrowFunctionExpression", 1, { parserOptions: { ecmaVersion: 6 } }),

